File: app/src/main/java/com/example/win/easy/viewmodel/SongListViewModelImpl.java
package com.example.win.easy.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import com.example.win.easy.repository.db.data_object.SongDO;
import com.example.win.easy.repository.db.data_object.SongListDO;
import com.example.win.easy.repository.repo.Repo;
import com.example.win.easy.value_object.SongListVO;
import com.example.win.easy.value_object.SongVO;
import com.example.win.easy.value_object.VOUtil;
import com.example.win.easy.view.main.SongListToCreateAlreadyExistLocallyException;

import java.util.ArrayList;
import java.util.List;

public class SongListViewModelImpl extends SongListViewModel {

    private Repo repo;
    private VOUtil voUtil;
    private LiveData<List<SongListVO>> allSongList;

    public SongListViewModelImpl(Repo repo, VOUtil voUtil){
        this.repo=repo;
        this.voUtil=voUtil;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public LiveData<List<SongListVO>> loadAll() {
        //懒惰加载
        if (allSongList==null)
            allSongList= initAllSongListLiveData();
        return allSongList;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public List<SongVO> songsOf(SongListVO songListVO) {
        //获取，避免空值，转化类型
        List<SongDO> songsInRepo=repo.songsOf(voUtil.toDO(songListVO));
        songsInRepo=avoidNullSongs(songsInRepo);
        return toSongsVO(songsInRepo);
    }

    @Override
    public List<SongVO> songsNotIn(SongListVO songListVO) {
        return null;
    }

    @Override
    public void addSongsTo(List<? extends SongVO> songVOs, SongListVO songListVO) {

    }

    @Override
    public void create(SongListVO songListVO) throws SongListToCreateAlreadyExistLocallyException {

    }

    private LiveData<List<SongListVO>> initAllSongListLiveData(){

        LiveData<List<SongListDO>> allSongListDOLiveData=repo.getAllSongList();

        //空值避免，类型转换
        return Transformations.map(allSongListDOLiveData,allSongListFromRepo->{
            List<SongListDO> allSongListDO= avoidNullSongLists(allSongListFromRepo);
            return toAllSongListVO(allSongListDO);
        });
    }

    private List<SongListDO> avoidNullSongLists(List<SongListDO> songListDOs){
        return songListDOs==null
                ? new ArrayList<>()
                :songListDOs;
    }

    private List<SongListVO> toAllSongListVO(List<SongListDO> songListDOs){
        List<SongListVO> allSongListVO=new ArrayList<>();
        for (SongListDO songListDO:songListDOs)
            allSongListVO.add(voUtil.toVO(songListDO));
        return allSongListVO;
    }

    private List<SongDO> avoidNullSongs(List<SongDO> songDOs){
        return songDOs==null
                ?new ArrayList<>()
                :songDOs;
    }

    private List<SongVO> toSongsVO(List<SongDO> songDOs){
        List<SongVO> songsVO=new ArrayList<>();
        for (SongDO songDO:songDOs)
            songsVO.add(voUtil.toVO(songDO));
        return songsVO;
    }
}

File: app/src/main/java/com/example/win/easy/viewmodel/SongListViewModelImpl.kt
package com.example.win.easy.viewmodel

import androidx.lifecycle.*
import com.example.win.easy.enumeration.DataSource
import com.example.win.easy.exception.SongListToCreateAlreadyExistLocallyException
import com.example.win.easy.repository.SongListRepository
import com.example.win.easy.db.SongDO
import com.example.win.easy.db.SongListDO
import com.example.win.easy.repository.SongRepository
import kotlinx.coroutines.launch

class SongListViewModelImpl(
        private val songRepository: SongRepository,
        private val songListRepository: SongListRepository) : SongListViewModel() {

    private val allSongList= songListRepository.allLiveData
    private lateinit var songsInList: MutableLiveData<List<SongDO>?>

    private val _spinner=MutableLiveData<Boolean>()
    val spinner: LiveData<Boolean>
        get() = _spinner

    private val _snackbar=MutableLiveData<String?>()
    val snackbar: LiveData<String?>
        get()= _snackbar


    /**
     * 尝试获取所有歌单：
     * 1. 未超时：获取本地+网络
     * 2. 超时：仅获取本地
     */
    override fun loadAll(): LiveData<List<SongListDO>>{
        launchLoading {
            val newSongLists=songListRepository.loadOnline()!!
            val oldSongLists=songListRepository.onlineLiveData.value?.toList()?: emptyList()
            viewModelScope.launch {
                songListRepository.syncSongLists(oldSongLists,newSongLists)
            }
            songListRepository.onlineLiveData.value=newSongLists
        }
        return allSongList
    }

    /**
     * 加载歌单内的歌曲
     */
    override fun loadSongsIn(songListDO: SongListDO):LiveData<List<SongDO>?>{
        return if(songListDO.source==DataSource.Local)
            songListRepository.loadSongsIn(songListDO)
        else{
            songsInList=MutableLiveData<List<SongDO>?>()
            refreshOnlineSongList(songListDO)
            songsInList
        }
    }

    @Throws(SongListToCreateAlreadyExistLocallyException::class)
    override fun insert(songListDO: SongListDO) {
        allSongList.value?.let {
            for (existedSongList in it)
                if (existedSongList.name==songListDO.name&&existedSongList.source==DataSource.Local)
                    throw SongListToCreateAlreadyExistLocallyException()
        }
        viewModelScope.launch {
            songListRepository.insert(songListDO)
        }
    }

    /**
     * 刷新歌单内的歌曲（仅网络歌单应当使用）
     */
    private fun refreshOnlineSongList(songListDO: SongListDO){
        launchLoading {
            songsInList.value=songListRepository.loadOnlineSongsIn(songListDO)
            viewModelScope.launch {
                songRepository.syncSongsInSongList(songListDO,songsInList.value!!)
            }
        }
    }

    /**
     * 统一的loading状态控制
     * @param block 执行时在viewModelScope中，它将执行限制在UI线程中，因此对viewmodel的赋值可以直接set
     */
    private fun launchLoading(block: suspend ()->Unit)=viewModelScope.launch{
        try {
            _spinner.postValue(true)
            block()
        }catch (t: Throwable){
            _snackbar.postValue(t.message)
        }finally {
            _spinner.postValue(false)
        }
    }

    override fun launchLoadSongListsBySong(songDO: SongDO, block: (songs: List<SongListDO>) -> Unit){
        viewModelScope.launch {
            block(songListRepository.loadSongListBySong(songDO))
        }
    }
}